core/src/sumeragi/fault.rs

```diff
@@ -1305,500 +872,511 @@ pub fn run_sumeragi_main_loop<F>(
-                            // sumeragi.txs_awaiting_created_block.clear(); TODO: Figure out what this is for
+                        let block = {
+                            let span_for_sumeragi_proxy_tail_block_validate =
+                                span!(Level::TRACE, "Sumeragi Validating Peer Validate block.");
+                            let _enter_for_sumeragi_proxy_tail_block_validate =
+                                span_for_sumeragi_proxy_tail_block_validate.enter();
 
-                            let voting_block = VotingBlock::new(block.clone());
-                            let voting_block_hash = voting_block.block.hash();
-                            voting_block_option = Some(voting_block);
+                            let block = block.revalidate(
+                                &sumeragi.transaction_validator,
+                                &state_machine_guard.wsv,
+                            );
+                            block
+                        };
 
-                            // TODO: Do commit countdown.
+                        let valid_signatures = state_machine_guard
+                            .current_topology
+                            .filter_signatures_by_roles(
+                                &[Role::ValidatingPeer, Role::Leader],
+                                block.verified_signatures(),
+                            );
+                        for sig in &valid_signatures {
+                            block_signature_acc.push((block.hash(), sig.clone()));
                         }
-                        Message::BlockSigned(block_signed) => {
-                            println!("block signed message");
-                            let block = block_signed.block;
-                            let block_hash = block.hash();
 
-                            if voting_block_option.is_some()
-                                && block_hash != voting_block_option.as_ref().unwrap().block.hash()
-                            {
-                                println!("block signed is not relevant block");
-                                continue;
-                            }
+                        let voting_block = VotingBlock::new(block.clone());
+                        voting_block_option = Some(voting_block);
 
-                            // I don't think we update the topology here. That ship has sailed.
+                        instant_at_which_we_should_have_committed =
+                            Instant::now() + sumeragi.commit_time;
+                    }
+                    Message::BlockSigned(block_signed) => {
+                        let block = block_signed.block;
+                        let block_hash = block.hash();
 
-                            let network_topology = &state_machine_guard.current_topology;
+                        if voting_block_option.is_some()
+                            && block_hash != voting_block_option.as_ref().unwrap().block.hash()
```

Comment:
Maybe?

```rust
if Some(block_hash) != voting_block_option.map(|block| block.block.hash()) {
    ...
}
```

Reply:
This one I think is a bit too much. It's too *clever*.